File: app/src/ipc/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/**
 * Bidirectional event bus type catalogue.
 * Add new event names + payload types as features land.
 * Mirrors the payload structs in `app/src-tauri/src/events.rs`. Each entry here should have a corresponding `#[derive(Serialize, Clone)] struct` on the Rust side.
 */
export interface WaterEventPayloads {
  "bus:ping": { tick: number };
  "sidecar:status": {
    status: "loading" | "ready" | "error";
    detail: string | null;
  };
  "provider:status": {
    provider_id: string;
    ok: boolean;
    error: string | null;
  };
  "typing:telemetry": {
    idle_for_ms: number;
    cursor_classification: "at_sentence_end" | "at_paragraph_end" | "mid_sentence";
    block_id: string;
    recent_word_delta: number;
    structural_inflection: "new_scene" | "new_chapter" | "pov_change" | "location_change" | "none";
    /**
     * Text of the current block; only populated on idle pulses
     * (`idle_for_ms >= 3000`). `null` during typing bursts (5 Hz cap)
     * to keep the renderer→core wire size small. Consumed by
     * `character_dissonance` via the orchestrator's `AnalysisSnapshot`.
     */
    last_block_text: string | null;
  };
  "pill:emerged": {
    pill_id: string;
    speaker_id: string;
    hue_token: string;
    text: string;
    block_target_id: string | null;
    trigger_id: string;
  };
  "pill:dismissed": { pill_id: string };
  "pill:evicted": { pill_id: string };
  "pill:pinned": import("../pill/types").Pill;
  "pill:unpinned": { pill_id: string };
  "bouquet:ready": {
    parent_pill_id: string;
    items: Array<{
      sub_pill_id: string;
      angle: "feel" | "notice" | "wonder";
      text: string;
    }>;
  };
  /**
   * M5: emitted by the orchestrator after a heat-metric recompute
   * lands. The renderer's HeatmapStrip subscribes and re-fetches
   * `heat_read` for the matching scene; small payload so we don't
   * spam every metric track. Fires on every autosave.
   */
  "heat:updated": { scene_id: string };
  /**
   * Phase 4: emitted when the rabbit_fan_4 LLM call returns and the
   * four children have been persisted. `parent_id` is the rabbit
   * thought that was deepened (root or interior).
   */
  "deepen:ready": {
    parent_id: string;
    children: Array<{
      id: string;
      direction: "closer" | "wider" | "opposite" | "deeper" | "root" | "";
      text: string;
    }>;
  };
  /** Phase 4: emitted when LLM dispatch or persistence failed. */
  "deepen:failed": { parent_id: string; reason: string };
  /**
   * Phase 5: emitted after a diagnostic-engine run upserts editor
   * pills for a scene. Renderer's diagnostics tab + inline-underline
   * plugin re-fetch via `editor_pills_list` on receipt.
   */
  "editor_pills:updated": { scene_id: string; count: number };
}

export type WaterEventName = keyof WaterEventPayloads;

export async function onWaterEvent<K extends WaterEventName>(
  name: K,
  cb: (payload: WaterEventPayloads[K]) => void,
): Promise<UnlistenFn> {
  return listen<WaterEventPayloads[K]>(name, (e) => cb(e.payload));
}

File: app/src-tauri/src/events.rs
//! Typed event-bus helpers for renderer↔core communication. Mirrors
//! `app/src/ipc/events.ts::WaterEventPayloads`. Add variants as features land.

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter};

/// Emit a typed event to the renderer. Returns Ok on success; errors are
/// usually "no window yet" during early boot — log and continue.
pub fn emit<T: Serialize + Clone>(
    app: &AppHandle,
    event_name: &str,
    payload: T,
) -> Result<(), tauri::Error> {
    app.emit(event_name, payload)
}

#[derive(Serialize, Clone)]
pub struct BusPing {
    pub tick: u64,
}

/// Payload for the `sidecar:status` event. Mirrors
/// `WaterEventPayloads["sidecar:status"]` in `app/src/ipc/events.ts`.
/// `status` is one of `"loading" | "ready" | "error"` — the renderer is
/// allowed to assume the closed set.
#[derive(Serialize, Clone)]
pub struct SidecarStatusPayload {
    pub status: String,
    pub detail: Option<String>,
}

/// Payload for the `provider:status` event. Mirrors
/// `WaterEventPayloads["provider:status"]` in `app/src/ipc/events.ts`.
/// Emitted whenever an `LlmRouter` provider transitions success/failure
/// state during a `generate_bouquet` call (currently exercised via
/// `provider_test`).
#[derive(Serialize, Clone)]
pub struct ProviderStatusPayload {
    pub provider_id: String,
    pub ok: bool,
    pub error: Option<String>,
}

/// Payload for the `typing:telemetry` event. Mirrors
/// `WaterEventPayloads["typing:telemetry"]` in `app/src/ipc/events.ts`.
/// Strings are closed sets — see the TS side for the allowed values.
#[derive(Serialize, Deserialize, Clone)]
pub struct TypingTelemetryPayload {
    pub idle_for_ms: u64,
    pub cursor_classification: String, // "at_sentence_end" | "at_paragraph_end" | "mid_sentence"
    pub block_id: String,
    pub recent_word_delta: i32,
    pub structural_inflection: String, // "new_scene" | "new_chapter" | "pov_change" | "location_change" | "none"
    /// Text of the current block; only populated on idle pulses
    /// (`idle_for_ms >= 3000`). `None` during typing bursts (5 Hz cap) so
    /// the renderer→core wire size stays small during live typing.
    /// Consumed by the orchestrator to populate
    /// `AnalysisSnapshot.last_block_text`, which `character_dissonance`
    /// gates against.
    pub last_block_text: Option<String>,
}
